Rework PyObject preservation

colesbury opened the pull request:
Make the PyObject preservation scheme thread-safe with free threaded (nogil) Python. The general idea is:

* Python Tensor and Storage objects always hold a strong reference to their underlying c10 object
* c10 objects hold a strong reference to their Python objects if there's at least one other reference to the c10 object

This is implemented in `intrusive_ptr`:

* The top most bit (`kHasPyObject`) from the weakref count is now used to indicate if the `intrusive_ptr_target` has an associated PyObject. So `kHasPyObject` is one bit, the weakref count is now 31 bits and the strong refcount remains 32 bits.
* When the reference count increases from one to two and `kHasPyObject` is set, we incref the associated Python object to ensure that it's kept alive. 
* When the reference count decreases from two to one (i.e., there are no C++ reference to the `intrusive_ptr_target` other than from the Python object), we decre the associated Python object to break the cycle.

Other benefits:

* We can delete a lot of the copypasta from Python internal `subtype_dealloc`
* This fixes the weakref and GC bugs we had in the previous scheme. Python weakrefs on Tensors and Storages should just work as expected now.

Risks:

* Extra branch for reference count operations on `intrusive_ptr<TensorImpl>`, `intrusive_ptr<StorageImpl>`, and the generic `intrusive_ptr<intrusive_ptr_target>` even when we're not using Python.
* It's a big change

colesbury committed:
[WIP] Rework PyObject preservation

colesbury committed:
Lint

colesbury committed:
Clean up

colesbury committed:
Delete THPStorageMetaType

colesbury committed:
Remove dead code

colesbury committed:
Fix Tensor traverse and THPStorageType.tp_base

colesbury committed:
Tensor and Storage Python objects always own their cdata

colesbury committed:
lint

swolchok commented on c10/util/intrusive_ptr.h:
```diff
@@ -33,6 +33,8 @@ constexpr uint64_t kImpracticallyHugeWeakReferenceCount =
 constexpr uint64_t kWeakReferenceCountOne = (kReferenceCountOne << 32);
 constexpr uint64_t kUniqueRef = (kReferenceCountOne | kWeakReferenceCountOne);
+// Indicates whether the object has a PyObject wrapper.
+constexpr uint64_t kHasPyObject = (uint64_t(1) << 63);
```
can/should we define a compile-time check to rule out some types T such that `intrusive_ptr<T>` cannot possibly have a pyobject and 1) save the extra instructions for those types and 2) check this compile-time assumption (ideally at compile time, but runtime is OK) by failing if the new pyobject stuff is called? I imagine that those types exist (do we ever back e.g. c10::List with a PyObject?) and would at least appreciate the code size savings

swolchok replied:
oops, I realize now that I should have suggested that instead we allowlist those types that require PyObject support rather than ruling out specific types.

swolchok replied:
> But in a few places we use the generic intrusive_ptr<intrusive_ptr_target> and I think that would end up (incorrectly) ignoring the PyObject wrappers when the actual target is a Tensor/Storage.

Couldn't we solve that by opting intrusive_ptr_target into the has_pyobject behavior? this shouldn't affect subclasses; instantiating TargetTraits<T> shouldn't care about inheritance

colesbury committed:
Move is_tensor_stealable to header

Avoids some linker issues with missing functions.

colesbury committed:
Use template magic

Only check the "has_pyobject" bit if the target type supports PyObject
wrappers.

colesbury committed:
Remove dead code

colesbury committed:
Suppress valgrind false positive

colesbury committed:
Friendship attempt take two

colesbury committed:
format

colesbury committed:
Mark self as maybe_unused for GCC 9

colesbury committed:
Atomic initialization and clean up

colesbury committed:
Clean up

colesbury committed:
Fix PyObjectSlot forward definition

colesbury committed:
Remove unused has_pyobj_slot

colesbury committed:
Combine THPVariable_NewFresh into THPVariable_WrapWithType

colesbury committed:
Fix lint

colesbury committed:
Merge branch 'main' into pyobjectslot

colesbury committed:
Add Tensor.untyped_storage() thread safety test

colesbury committed:
Fix remaining weak_intrusive_ptr::lock thread-safety

colesbury committed:
Clean up intrusive_ptr

Fix NVCC + MSVC bug with std::atomic_thread_fence call in
intrusive_ptr.h

colesbury committed:
Clear pyobject slot during dealloc

colesbury committed:
Use C (not C++) in python_compat.h

colesbury committed:
Fix refcount issue with THPModule_swap_tensor_impl

colesbury committed:
Merge branch 'main' into pyobjectslot

colesbury committed:
Merge remote-tracking branch 'origin/main' into pyobjectslot

colesbury commented on c10/core/StorageImpl.h:
```diff
@@ -105,6 +105,30 @@ struct C10_API StorageImpl : public c10::intrusive_ptr_target {
+    // do an acquire fence to ensure that the kHasPyObject bit was
+    // observed before the load of the PyObject* below.
+    // NB: This is a no-op on x86/x86-64
+    std::atomic_thread_fence(std::memory_order_acquire);
```
Benchmark: https://gist.github.com/colesbury/27ed5b8bbe01dc2eabbfe5285d0aa1d7

Apple M4: ~0.38 ns (so maybe ~1-2 cycles)
AWS t4g: ~2.85 ns (so maybe 8 cycles)

In my experience, all the non Apple aarch64 chips have pretty similar performance for memory barriers (with Apple's chips performing much better). 

We could also get away with a reload `combined_refcount_.load(std::memory_order_acquire)`. That compiles to `ldar` instead of `dmb ish` and is a bit faster. It's just a bit awkward in the code, and GCC won't optimize the atomic load away on x86-64 like it will the fence:

Apple M4: ~0.22 ns (so maybe ~1 cycle)
AWS t4g: ~0.82 ns (so maybe ~2-3 cycles)

mcfi replied:
I think a reload of the counter with acquire seems to be the best choice for arm64 since we already loaded the counter so the reload (ldar if targeting Armv8.0 and ldapr if targeting Armv8.2 + LRCPC) should hit L1 cache and be the fastest compared to a dmb fence on Arm CPUs.

For x86, the acquire fence is a nop, so it might be good to have the following defined in intrusive_ptr.h and used everywhere.
#ifdef x86
#define acquire_fence std::atomic_thread_fence(std::memory_order_acquire)
#elif arm64
#define acquire_fence combined_refcount_.load(std::memory_order_acquire)
#endif

colesbury replied:
I've updated the benchmark with:

fetch_add

```c++
x.fetch_add(1, std::memory_order_relaxed);
```

fetch_add + reload

```c++
x.fetch_add(1, std::memory_order_relaxed);
x.load(std::memory_order_acquire);
```

Apple M4: 3.5 ns/iter
AWS t4g: 8.2 ns/iter

fetch_add + acquire fence (`dmb ishld`)

```c++
x.fetch_add(1, std::memory_order_relaxed);
std::atomic_thread_fence(std::memory_order_acquire);
``` 

The fence is no slower, sometimes faster:

|          | fetch_add | fetch_add + reload | fetch_add + fence |
|----------|-----------|--------------------|-------------------|
| Apple M4 | 1.6 ns    | 3.6 ns             | 1.6 ns            |
| AWS t4g  | 5.2 ns    | 8.2 ns             | 8.0 ns            |

(Also I remembered to compile with `-march=armv8.3-a` for the aws graviton this time, although it doesn't make much of a difference)

mcfi replied:
ok, thanks for the new microbenchmark data. I'm fine either way.

colesbury committed:
Move functions to .cpp file in attempt to decrease binary size

colesbury committed:
Merge remote-tracking branch 'origin/main' into pyobjectslot

albanD commented on test/test_torch.py:
```diff
@@ -10407,6 +10402,31 @@ def callback(w):
+
+                # Check that all the storages returned were the same
+                for future in futures:
+                    self.assertEqual(future.result(), id(tensor.untyped_storage()))
```
nit: given that we don't keep the storage object alive, could this be passing by mistake because it re-uses the same memory space for all the storage objects one after the other?
Should we use weakrefs here to check for identity?

colesbury replied:
Yeah, I can do that

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -180,7 +180,9 @@ struct TORCH_API AccumulateGrad : public Node {
-          at::caching::adjusted_use_count(new_grad) <= num_expected_refs &&
+          impl::is_tensor_stealable(
+              new_grad,
+              num_expected_refs + at::caching::is_cached_tensor(new_grad)) &&
```
Why the addition of is_cached_tensor here?

albanD commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -292,6 +254,33 @@ void ConcretePyInterpreterVTable::incref(PyObject* pyobj) const {
+  if (!pyobj) {
+    return false;
+  }
+#if IS_PYTHON_3_14_PLUS
```
Not needed given the api compat addition

albanD reviewed (commented):
Very nice cleanup!

colesbury committed:
Move StorageImpl overloads to .cpp file

colesbury committed:
Merge remote-tracking branch 'origin/main' into pyobjectslot

colesbury commented on torch/csrc/utils/python_compat.h:
```diff
@@ -33,12 +33,67 @@ static inline int PyCode_GetNFreevars(PyCodeObject* code) {
-// NOLINTNEXTLINE(readability-redundant-declaration)
-PyAPI_FUNC(void) _PyWeakref_ClearRef(PyWeakReference* self);
+#if !IS_PYTHON_3_14_PLUS
+static inline int PyUnstable_TryIncRef(PyObject* obj) {
```
Yes, removed now

colesbury committed:
Just use PyUnstable_TryIncRef in PyInterpreter.cpp

colesbury commented on torch/csrc/autograd/utils/grad_layout_contract.h:
```diff
@@ -65,7 +65,9 @@ inline at::Tensor clone_obey_contract(
-                             variable.options().memory_format(std::nullopt))
+                             variable.options()
+                                 .memory_format(std::nullopt)
+                                 .dtype(new_grad.dtype()))
```
This fixes a semi-related bug. I can pull it out into a separate PR if you prefer. Basically `grad_dtype` was ignored if we copied instead of stole.

albanD replied:
Ho right.
FYI @soulitzer

albanD reviewed (approved):
Looks good.
Thanks for the simplification!

colesbury committed:
Optimize THPVariable_Wrap for fresh Tensors

colesbury committed:
Use weakref in test_storage_thread_safety

colesbury committed:
Clear pyobj_slot in destructor

colesbury committed:
Add PyUnstable_EnableTryIncRef calls

colesbury committed:
Merge branch 'main' into pyobjectslot

colesbury committed:
Add a thread safety test for Tensor wrapping

Add a comment for the PyUnstable_EnableTryIncRef calls.

albanD reviewed (approved):
SGTM
